frontend/public/components/hpa.tsx

```diff
@@ -31,8 +34,118 @@ const menuActions = [
   ...common,
 ];
 
-export const HorizontalPodAutoscalersDetails: React.SFC<HorizontalPodAutoscalersDetailsProps> = ({obj: hpa}) => {
-  return <div className="co-m-pane__body">
+const MetricsRow: React.SFC<MetricsRowProps> = ({type, current, target}) => <div className="row">
+  <div className="col-xs-6">
+    {type}
+  </div>
+  <div className="col-xs-3">
+    {current || '-'}
+  </div>
+  <div className="col-xs-3">
+    {target || '-'}
+  </div>
+</div>;
+
+const externalRow = (metric, current, key) => {
+  const { external } = metric;
+  const type = external.metricName;
+  // TODO: show metric selector for external metrics?
+  const currentValue = external.targetAverageValue
+    ? _.get(current, 'object.currentAverageValue')
+    : _.get(current, 'object.currentValue');
+  const targetValue = external.targetAverageValue
+    ? external.targetAverageValue
+    : external.targetValue;
+
+  return <MetricsRow key={key} type={type} current={currentValue} target={targetValue} />;
+};
+
+const objectRow = (metric, current, ns, key) => {
+  const { object } = metric;
+  const type = <React.Fragment>
+    {object.metricName} on
+    <ResourceLink kind={object.target.kind} name={object.target.name} namespace={ns} title={object.target.name} />
+  </React.Fragment>;
+  const currentValue = _.get(current, 'object.currentValue');
+  const targetValue = object.targetValue;
+
+  return <MetricsRow key={key} type={type} current={currentValue} target={targetValue} />;
+};
+
+const podRow = (metric, current, key) => {
+  const { pods } = metric;
+  const type = `${pods.metricName} on pods`;
+  const currentValue = _.get(current, 'pods.currentAverageValue');
+  const targetValue = pods.targetAverageValue;
+
+  return <MetricsRow key={key} type={type} current={currentValue} target={targetValue} />;
+};
+
+const getUtilization = (current, metricUtilization) => {
```

Comment:
This name is confusing:

* It's not clear if it gets target average utilization or current average utilization
* This function sometimes returns current average value instead of a utilization percentage, so it's not always returning "utilization"
* It's not clear that this is only for "Resource" type metrics

I would suggest calling it something like `getCurrentResourceUsage` as written.

But I think this would be better:

```jsx
const getCurrentResourceUtilization = currentMetric => {
  const currentUtilization = _.get(current, 'resource.currentAverageUtilization');
  const currentAverageValue = _.get(current, 'resource.currentAverageValue');
  // Fallback to currentAverageValue if utilization is not set.
  return _.isNumber(currentUtilization) ? `${currentUtilization}% (${currentAverageValue})`  : currentAverageValue;
};
```

Then only call it below when we want utilization:

```js
const currentValue = metricUtilization ? getUtilization(current) : _.get(current, 'resource.currentAverageValue');
const targetValue = metricUtilization ? `${metricUtilization}%` : metric.resource.targetAverageValue;
```

If we're going to check the same condition twice, it's better if the checks are side-by-side so that they don't get out of sync. The proposed refactor puts them in separate methods.

If we refactor code to pull out methods, it's important that the helpers do one clear thing and have good names.